### DISCUSSION TITLE
Help Needed

### DISCUSSION THREAD

**Discussion by SpencerHD2020** [nlohmann/json#3607] [Q&A] [answered]
I am getting a compiler error (G++) and I am not sure why. I will paste the compiler output at the bottom but it seems to be complaining about the line where I attempt json::parse on an item within an array from a JSON string that I parsed earlier in the code. I can run a for loop on the array and print out a JSON string fine for each item (syntax: "[{"foo": "bar", ...}]"). I will paste the code below any help is greatly appreciated! I have tried passing the direct array ref to the parse method as well as a pointer to it and saving it in an auto variable first but all give the same error. I added some "bold comments" to the code below to show a similar method call working and where the failure was introduced.

// My Code

void myFunction(std::string query) { 
    // Establish Client Connection
    httplib::Client cli("http://localhost:4000");
    // GET Request
    httplib::Result res = cli.Get("/searchCards?q=" + query);
    //std::cout << res->body << std::endl;
    // Save JSON Body as String
    auto jsonString = res->body;

    //**THIS PARSE IS WORKING FINE**
    // Parse JSON into JSON Object
    json parsedJSON = json::parse(jsonString);
   

 // Access Values
    int status = parsedJSON.value("status", -1);
    int dataLength = parsedJSON.value("length", -1);
    auto cardArr = parsedJSON["data"];
    if (status == 0 && dataLength > 0) {
        // We should have valid data
        for (int i = 0; i < dataLength; ++i) {
            // This is working but I imagine these are strings
            //std::cout << "******************************" << std::endl;
            //std::cout << cardArr[i] << std::endl;
            //std::cout << "******************************" << std::endl;

            auto tempCardString = cardArr[i];
            // **THIS LINE OF CODE IS CAUSING THE COMPILER ERRORS**
            auto tempCardObj = json::parse(tempCardString);

            //std::cout << std::setw(4) << tempCardObj << std::endl;
            break;
        }
    }
    else {
        std::cout << "Invalid Data Received!" << std::endl;
        std::cout << "Data Length: " << dataLength << std::endl;
        std::cout << "Status: " << status << std::endl;
    }
}

// GCC Output
./json/include/nlohmann/detail/input/input_adapters.hpp:330:9: error: implicit instantiation of undefined template 'nlohmann::detail::wide_string_input_helper<nlohmann::detail::iterator_input_adapter<nlohmann::detail::iter_impl<const nlohmann::basic_json<>>>, 16>'
        wide_string_input_helper<BaseInputAdapter, T>::fill_buffer(base_adapter, utf8_bytes, utf8_bytes_index, utf8_bytes_filled);
        ^
./json/include/nlohmann/detail/input/input_adapters.hpp:312:13: note: in instantiation of function template specialization 'nlohmann::detail::wide_string_input_adapter<nlohmann::detail::iterator_input_adapter<nlohmann::detail::iter_impl<const nlohmann::basic_json<>>>, nlohmann::basic_json<>>::fill_buffer<16>' requested here
            fill_buffer<sizeof(WideCharType)>();
            ^
./json/include/nlohmann/detail/input/lexer.hpp:1349:26: note: in instantiation of member function 'nlohmann::detail::wide_string_input_adapter<nlohmann::detail::iterator_input_adapter<nlohmann::detail::iter_impl<const nlohmann::basic_json<>>>, nlohmann::basic_json<>>::get_character' requested here
            current = ia.get_character();
                         ^
./json/include/nlohmann/detail/input/lexer.hpp:1488:13: note: in instantiation of member function 'nlohmann::detail::lexer<nlohmann::basic_json<>, nlohmann::detail::wide_string_input_adapter<nlohmann::detail::iterator_input_adapter<nlohmann::detail::iter_impl<const nlohmann::basic_json<>>>, nlohmann::basic_json<>>>::get' requested here
        if (get() == 0xEF)
            ^
./json/include/nlohmann/detail/input/lexer.hpp:1512:48: note: in instantiation of member function 'nlohmann::detail::lexer<nlohmann::basic_json<>, nlohmann::detail::wide_string_input_adapter<nlohmann::detail::iterator_input_adapter<nlohmann::detail::iter_impl<const nlohmann::basic_json<>>>, nlohmann::basic_json<>>>::skip_bom' requested here
        if (position.chars_read_total == 0 && !skip_bom())
                                               ^
./json/include/nlohmann/detail/input/parser.hpp:464:37: note: in instantiation of member function 'nlohmann::detail::lexer<nlohmann::basic_json<>, nlohmann::detail::wide_string_input_adapter<nlohmann::detail::iterator_input_adapter<nlohmann::detail::iter_impl<const nlohmann::basic_json<>>>, nlohmann::basic_json<>>>::scan' requested here
        return last_token = m_lexer.scan();
                                    ^
./json/include/nlohmann/detail/input/parser.hpp:81:9: note: in instantiation of member function 'nlohmann::detail::parser<nlohmann::basic_json<>, nlohmann::detail::wide_string_input_adapter<nlohmann::detail::iterator_input_adapter<nlohmann::detail::iter_impl<const nlohmann::basic_json<>>>, nlohmann::basic_json<>>>::get_token' requested here
        get_token();
        ^
./json/include/nlohmann/json.hpp:148:16: note: in instantiation of member function 'nlohmann::detail::parser<nlohmann::basic_json<>, nlohmann::detail::wide_string_input_adapter<nlohmann::detail::iterator_input_adapter<nlohmann::detail::iter_impl<const nlohmann::basic_json<>>>, nlohmann::basic_json<>>>::parser' requested here
        return ::nlohmann::detail::parser<basic_json, InputAdapterType>(std::move(adapter),
               ^
./json/include/nlohmann/json.hpp:3971:9: note: in instantiation of function template specialization 'nlohmann::basic_json<>::parser<nlohmann::detail::wide_string_input_adapter<nlohmann::detail::iterator_input_adapter<nlohmann::detail::iter_impl<const nlohmann::basic_json<>>>, nlohmann::basic_json<>>>' requested here
        parser(detail::input_adapter(std::forward<InputType>(i)), cb, allow_exceptions, ignore_comments).parse(true, result);
        ^
main.cpp:32:38: note: in instantiation of function template specialization 'nlohmann::basic_json<>::parse<nlohmann::basic_json<> &>' requested here
            auto tempCardObj = json::parse(tempCardString);
                                     ^
./json/include/nlohmann/detail/input/input_adapters.hpp:175:8: note: template is declared here
struct wide_string_input_helper;
       ^
./json/include/nlohmann/detail/input/input_adapters.hpp:330:54: error: incomplete definition of type 'nlohmann::detail::wide_string_input_helper<nlohmann::detail::iterator_input_adapter<nlohmann::detail::iter_impl<const nlohmann::basic_json<>>>, 16>'
        wide_string_input_helper<BaseInputAdapter, T>::fill_buffer(base_adapter, utf8_bytes, utf8_bytes_index, utf8_bytes_filled);
        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~
2 errors generated.

**Comment by falbrechtskirchinger** [comment, marked as answer]
`tempCardString` – despite its name – isn't a string. Try

```cpp
auto tempCardObj = json::parse(tempCardString.get_ref<const std::string &>());
```

or

```cpp
auto tempCardObj = json::parse(tempCardString.get<std::string>());
```

instead.

**Reply by SpencerHD2020** [reply]
Thank You! That still threw an error but let me know that the value was already an object thus that data was parsed in the first call. I am newer to C++, coming over from higher level languages, so I greatly appreciate the quick response!
